packages/react-components/react-combobox/src/components/Listbox/useListbox.ts

```diff
@@ -27,66 +31,80 @@ import type { ListboxProps, ListboxState } from './Listbox.types';
+    controller,
+  } = useActiveDescendant<HTMLInputElement, HTMLDivElement>({
+    matchOption: el => el.classList.contains(optionClassNames.root),
+  });
 
-  const [activeOption, setActiveOption] = React.useState<OptionValue | undefined>();
+  const activeDescendantContext = useActiveDescendantContext();
+  const activeDescendantController = useHasParentActiveDescendantContext()
+    ? activeDescendantContext.controller
+    : controller;
 
-  // track whether keyboard focus outline should be shown
-  // tabster/keyborg doesn't work here, since the actual keyboard focus target doesn't move
-  const [focusVisible, setFocusVisible] = React.useState(false);
+  const { clearSelection, selectedOptions, selectOption } = useSelection(props);
 
   const onKeyDown = (event: React.KeyboardEvent<HTMLElement>) => {
     const action = getDropdownActionFromKey(event, { open: true });
-    const maxIndex = getCount() - 1;
-    const activeIndex = activeOption ? getIndexOfId(activeOption.id) : -1;
-    let newIndex = activeIndex;
+    const activeOptionId = activeDescendantController.active();
+    const activeOption = activeOptionId ? getOptionById(activeOptionId) : null;
 
     switch (action) {
+      case 'Next':
+        if (activeOption) {
+          activeDescendantController.next();
+        } else {
+          activeDescendantController.first();
+        }
+        break;
+      case 'Previous':
+        if (activeOption) {
+          activeDescendantController.prev();
+        } else {
+          activeDescendantController.first();
+        }
+        break;
+      case 'PageUp':
+      case 'First':
+        activeDescendantController.first();
+        break;
+      case 'PageDown':
+      case 'Last':
+        activeDescendantController.last();
+        break;
       case 'Select':
       case 'CloseSelect':
         activeOption && selectOption(event, activeOption);
         break;
-      default:
-        newIndex = getIndexFromAction(action, activeIndex, maxIndex);
     }
-
-    if (newIndex !== activeIndex) {
-      // prevent default page scroll/keyboard action if the index changed
-      event.preventDefault();
-      setActiveOption(getOptionAtIndex(newIndex));
-      setFocusVisible(true);
-    }
-  };
-
-  const onMouseOver = (event: React.MouseEvent<HTMLElement>) => {
-    setFocusVisible(false);
   };
 
   // get state from parent combobox, if it exists
   const hasComboboxContext = useHasParentContext(ComboboxContext);
-  const comboboxActiveOption = useContextSelector(ComboboxContext, ctx => ctx.activeOption);
-  const comboboxFocusVisible = useContextSelector(ComboboxContext, ctx => ctx.focusVisible);
   const comboboxSelectedOptions = useContextSelector(ComboboxContext, ctx => ctx.selectedOptions);
   const comboboxSelectOption = useContextSelector(ComboboxContext, ctx => ctx.selectOption);
-  const comboboxSetActiveOption = useContextSelector(ComboboxContext, ctx => ctx.setActiveOption);
+
+  const noLongerUsed = {
```

Comment:
I suppose this object is there to not break api? it can be moved outside of this hook. Or do we want to keep it inside so the hook is pure?
Also what about some prefix like UNSAFE_noLongerUsed?